src/solvers/refinement/string_refinement.cpp

```diff
@@ -2091,23 +2091,26 @@ static void update_index_set(
 /// \param [in] expr: the expression to search
 /// \param [in] str: the string which must be indexed
 /// \param [in] qvar: the universal variable that must be in the index
-/// \return an index expression in `expr` on `str` containing `qvar`
-static exprt find_index(
-  const exprt &expr, const exprt &str, const symbol_exprt &qvar)
+/// \return an index expression in `expr` on `str` containing `qvar`. Returns
+///   an empty optional when `expr` does not contain `str`.
+static optionalt<exprt>
+find_index(const exprt &expr, const exprt &str, const symbol_exprt &qvar)
```

Comment:
Not sure how this is used, but I'm wondering whether a normalisation of the if expressions would help. Btw, what would the user of this function do if the expression contains two different indices, e.g. `cond?(str1[i]):(str1[i+1])`? If this cannot happen then probably a data_invariant would be justified.

Reply:
Maybe an normalisation of if expression would have helped, but I'm not sure this would be preserved by simply_expr for instance, so I prefer to make function able to handle that case for now.
I added an invariant for checking there are no two such indexes, one drawback is that this does a search whose result we don't really care about, but this should be pretty cheap.